httpobs/scanner/analyzer/headers.py

```diff
@@ -4,6 +4,50 @@
 from httpobs.scanner.analyzer.utils import is_hpkp_preloaded, is_hsts_preloaded, only_if_worse
 
 
+def __parse_csp(csp_string: str) -> dict:
+    """
+    Decompose the CSP; could probably do this in one step, but it's complicated enough
+    Should look like:
+    {
+      'default-src': {'\'none\''},
+      'object-src': {'\'none\''},
+      'script-src': {'https://mozilla.org', '\'unsafe-inline\''},
+      'style-src': {'\'self\', 'https://mozilla.org'},
+      'upgrade-insecure-requests': {},
+    }
+    """
+
+    # Clean out all the junk
+    csp_string = csp_string.replace('\r', '').replace('\n', '').strip()
+
+    # So technically the shortest directive is img-src, so lets just assume that
+    # anything super short is invalid
+    if len(csp_string) < 6 or csp_string.isspace():
+        raise ValueError
+
+    # It's actually rather up in the air if CSP is case sensitive or not for directives, see:
+    # https://github.com/w3c/webappsec-csp/issues/236
+    # For now, we shall treat it as case-sensitive, since it's the safer thing to do, even though
+    # Firefox, Safari, and Edge all treat them as case-insensitive.
+    csp = {}
+
+    for entry in [directive.strip().split(maxsplit=1) for directive in csp_string.split(';') if directive]:
+        directive = entry[0]
+
+        # Technically the path part of any source is case-sensitive, but since we don't test
+        # any paths, we can cheat a little bit here
+        values = set([_.lower() for _ in entry[-1].split()]) if len(entry) > 1 else {'\'none\''}
+
+        # While technically valid in that you just use the first entry, we are saying that repeated
+        # directives are invalid so that people notice it
+        if directive in csp:
+            raise ValueError
```

Comment:
Since you're raising `ValueError`s for multiple reasons, I'd suggest doing one of two things:

1. Raising instances of `ValueError` with a `message`, e.g. `raise ValueError('Repeated directives are invalid')`
2. More complicated but probably more correct: create a series of custom error classes for each possible exception, so in the future you can explicitly except specific exception types.

```python
Class HTTPObservatoryException(Exception):
    Pass

Class CSPParsingException(HTTPObservatoryException):
    Pass

Class DirectiveLengthSception(CSPParsingException):
    Pass

Class RepeatedDirectiveException(CSPParsingException):
    Pass
```

(I'm on an iPad so first lines are capitalizing, but you get the idea.)

Reply:
Fixed, with the first option. I may eventually do the later, but the exceptions are all handled so they don't actually percolate upward in a useful manner, only the failure is.